lgc/tool/lgc/lgc.cpp

```diff
@@ -0,0 +1,259 @@
+ ***********************************************************************************************************************
+ */
+
+#include "lgc/LgcContext.h"
+#include "lgc/Pipeline.h"
+#include "llvm/AsmParser/Parser.h"
+#include "llvm/BinaryFormat/ELF.h"
+#include "llvm/IR/Verifier.h"
+#include "llvm/Support/CommandLine.h"
+#include "llvm/Support/Path.h"
+#include "llvm/Support/SourceMgr.h"
+
+using namespace lgc;
+using namespace llvm;
+
+// Input sources
+static cl::list<std::string> InFiles(cl::Positional, cl::OneOrMore, cl::ValueRequired,
+                                     cl::desc("Input file(s) (\"-\" for stdin)"));
+
+// -extract: extract a single module from a multi-module input file
+static cl::opt<unsigned> Extract("extract", cl::desc("Extract single module from multi-module input file"), cl::init(0),
+                                 cl::value_desc("1-based index of module to extract"));
+
+// -o: output filename
+static cl::opt<std::string> OutFileName("o", cl::desc("Output filename"),
+                                        cl::value_desc("filename (\"-\" for stdout)"));
+
+// -pal-abi-version: PAL pipeline ABI version to compile for (default is latest known)
+static cl::opt<unsigned> PalAbiVersion("pal-abi-version", cl::init(0xFFFFFFFF),
+                                       cl::desc("PAL pipeline version to compile for (default latest known)"));
+
+// =====================================================================================================================
+// Checks whether the input data is actually a ELF binary
+//
+// @param data : Input data to check
+static bool isElfBinary(StringRef data) {
+  bool isElfBin = false;
+  if (data.size() >= sizeof(ELF::Elf64_Ehdr)) {
+    auto pHeader = reinterpret_cast<const ELF::Elf64_Ehdr *>(data.data());
+    isElfBin = pHeader->checkMagic();
+  }
+  return isElfBin;
+}
+
+// =====================================================================================================================
+// Checks whether the output data is actually ISA assembler text
+//
+// @param data : Input data to check
+static bool isIsaText(StringRef data) {
+  // This is called by the lgc standalone tool to help distinguish between its three output types of ELF binary,
+  // LLVM IR assembler and ISA assembler. Here we use the fact that ISA assembler is the only one that starts
+  // with a tab character.
+  return data.startswith("\t");
+}
+// =====================================================================================================================
+// Main code of LGC standalone tool
+//
+// @param argc : Count of command-line arguments
+// @param argv : Command-line arguments
+int main(int argc, char **argv) {
+  const char *progName = sys::path::filename(argv[0]).data();
+  LLVMContext context;
+  LgcContext::initialize();
+  cl::ParseCommandLineOptions(argc, argv, "lgc: command-line tool for LGC, the LLPC middle-end compiler\n"
```

Comment:
You can use a raw string literal here to make this long string a bit more pretty.
But won't this help get printed out automatically based on the cl options? It seems very repetitive to specify these options 3 times: as `llvm::cl::opt`, in comments above, and in the help text again.

Reply:
lgc -help just prints all of the options, so the ones that are particularly relevant to lgc get lost. That's why I added this bit here. But I think I saw some other tool have a separate section for its own options at the top the other day, so maybe there is a generic mechanism for doing that that I haven't found yet.

Reply:
I figured out how to do this properly.